smart_card_connector_app/src/chrome-api-jstocxxtest.js

```diff
@@ -632,14 +651,168 @@ goog.exportSymbol('testChromeApiProviderToCpp', {
+        }
+
+        // Got the expected error code.
+        // Since result code is error, default values should be reported.
+        assertArrayEquals(
+            [
+              '', chrome.smartCardProviderPrivate.ConnectionState.ABSENT,
+              chrome.smartCardProviderPrivate.Protocol.UNDEFINED,
+              new ArrayBuffer(0), 'READER_UNAVAILABLE'
+            ],
+            saveArgs.allArgs);
+        break;
+      }
+    },
+
+    // Test that Status returns REMOVED_CARD after removing the card.
+    'testStatus_removedCard': async function() {
+      // Simulate the card removal.
+      setSimulatedDevices(/*devices=*/[
+        {'id': 123, 'type': SimulationConstants.GEMALTO_DEVICE_TYPE}
+      ]);
+
+      // Status should eventually return a REMOVED_CARD error.
+      for (let requestId = 111; ; requestId += 1) {
+        const saveArgs = new goog.testing.mockmatchers.SaveArgument();
+        expectReportStatusResult(
+            requestId, saveArgs, saveArgs, saveArgs, saveArgs, saveArgs);
+        await mockChromeApi
+            .dispatchEvent('onStatusRequested', requestId, sCardHandle)
+            .$waitAndVerify();
+
+        const resultCode = saveArgs.allArgs[saveArgs.allArgs.length - 1];
+        // If result code is still SUCCESS, continue.
+        if (resultCode == 'SUCCESS') {
+          assertArrayEquals(
+              [
+                SimulationConstants.GEMALTO_PC_TWIN_READER_PCSC_NAME0,
+                chrome.smartCardProviderPrivate.ConnectionState.NEGOTIABLE,
+                chrome.smartCardProviderPrivate.Protocol.T1,
+                SimulationConstants.COSMO_ID_70_ATR, 'SUCCESS'
+              ],
+              saveArgs.allArgs);
+          continue;
+        }
+
+        if (resultCode != 'REMOVED_CARD') {
+          goog.testing.asserts.raiseException(
+              `Unexpected result code: ${resultCode}`);
+        }
+
+        // Got the expected error code.
+        // Since result code is error, default values should be reported.
+        assertArrayEquals(
+            [
+              '', chrome.smartCardProviderPrivate.ConnectionState.ABSENT,
+              chrome.smartCardProviderPrivate.Protocol.UNDEFINED,
+              new ArrayBuffer(0), 'REMOVED_CARD'
+            ],
+            saveArgs.allArgs);
+        break;
+      }
+    },
+
+    // Test that GetAttrib succeeds for the `SCARD_ATTR_ATR_STRING` argument.
+    'testGetAttrib': async function() {
+      expectReportDataResult(
+          /*requestId=*/ 111, SimulationConstants.COSMO_ID_70_ATR, 'SUCCESS');
+
+      await mockChromeApi
+          .dispatchEvent(
+              'onGetAttribRequested', /*requestId=*/ 111, sCardHandle,
+              API.SCARD_ATTR_ATR_STRING)
+          .$waitAndVerify();
+    },
+
+    // Test that SetAttrib fails for unsupported attributes.
+    'testSetAttrib_error': async function() {
+    expectReportPlainResult(/*requestId=*/ 111, 'INVALID_PARAMETER');
```

Comment:
Formatting is off.

(On that topic, I spent half of the day trying to make a CI check on this - #1032 - but it's surprisingly difficult because Ubuntu's default clang-format apparently formats code differently than our regular developer machines' ones.)

Reply:
fixed